### PR TITLE
[Feature] Ability to extract files simultaneously

### PR DESCRIPTION
**Pull request by AlexSmirnov9107**
Fix or Feature to #10

### CODE REVIEW

File: android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java [hunk 1]
```diff
@@ -34,7 +34,11 @@ public void onMethodCall(@NonNull MethodCall call, @NonNull final Result result)
             waveformExtractor.start(new WaveformExtractor.OnProgressListener() {
                 @Override
                 public void onProgress(int progress) {
-                    invokeMethod("onProgress", progress);
+                     Map<String, Object> args = new HashMap();
+                    args.put("progress", progress);
+                    args.put("waveOutFile",waveOutPath);
```
Comment (ryanheise) [inline android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java:39]: Should put a space after the comma.

**Comment by ryanheise** [comment]
Going forward, I wonder whether it would be better to generate a unique ID for each instance rather than use the file name. The plan is to eventually include partial results of the waveform data in the progress so that eventually you might not need the file.

File: android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java [hunk 2]
```diff
@@ -19,52 +19,56 @@
...
-            Integer samplesPerPixel = call.argument("samplesPerPixel");
-            Integer pixelsPerSecond = call.argument("pixelsPerSecond");
-            WaveformExtractor waveformExtractor = new WaveformExtractor(audioInPath, waveOutPath, samplesPerPixel, pixelsPerSecond);
-            waveformExtractor.start(new WaveformExtractor.OnProgressListener() {
-                @Override
-                public void onProgress(int progress) {
-                     Map<String, Object> args = new HashMap();
-                    args.put("progress", progress);
-                    args.put("waveOutFile",waveOutPath);
+            case "extract":
```
Comment (ryanheise) [inline android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java:30]: Can you undo the change to this line and similar lines? Changes to the coding standards, such as whether cases should be indented or not, should be first discussed in an issue and then changed in its own pull request.

Comment (ryanheise) [reply android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java:30]: Just to clarify the current coding style, each switch case is considered as a label, not an executable line of code, and you don't indent labels. Of course there are other coding styles, but this is just to clarify what is currently being used in the Java code.

Comment (ryanheise) [reply android/src/main/java/com/ryanheise/just_waveform/JustWaveformPlugin.java:30]: FYI, the original Java coding conventions from SUN Microsystems: https://www.oracle.com/java/technologies/javase/codeconventions-statements.html#468

File: darwin/Classes/JustWaveformPlugin.m [hunk 3]
```diff
@@ -121,7 +121,7 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
             UInt32 scaledSampleIdx = 0;
             int progress = 0;
 
-            while (frameCount > 0) {
+            while (frameCount > 0 && progress < 100) {
```
Comment (ryanheise) [inline darwin/Classes/JustWaveformPlugin.m:124]: Just curious, what is the scenario where the second condition is required?

Comment (AlexSmirnov9107) [reply darwin/Classes/JustWaveformPlugin.m:124]: Sometimes the progress never become 100, kind of rare times. may be frame count cause the problem. Its difficult to find

Comment (ryanheise) [reply darwin/Classes/JustWaveformPlugin.m:125]: Your condition wouldn't help if the progress never became 100. Maybe you meant that frameCount never becomes 0?

**Comment by Faaatman** [comment]
This looks great! thank you guys for the great work.
@ryanheise if you think this is resource-intensive what do you think about adding a max number of concurrent jobs so that people could limit the number of files being extracted?
if this PR doesn't get merged, do you think we can at least throw an error if a job tries to run while another one is running?
I myself will be implementing a queue interface but I think an exception thrown is much needed.
I will be happy to help. :)

**Comment by ryanheise** [comment]
> @ryanheise if you think this is resource-intensive what do you think about adding a max number of concurrent jobs so that people could limit the number of files being extracted?

That is something that can/should be implemented by the app. (I like to keep things simple)

> if this PR doesn't get merged, do you think we can at least throw an error if a job tries to run while another one is running? I myself will be implementing a queue interface but I think an exception thrown is much needed. I will be happy to help. :)

At this stage my plan is to eventually merge this PR.

Before I do that, I'm just thinking about whether we want to have a unique ID for each instance generated by uuid or just use the file path as the unique ID.  (Again just thinking of simplicity.)

**Comment by Faaatman** [comment]
> Before I do that, I'm just thinking about whether we want to have a unique ID for each instance generated by uuid or just use the file path as the unique ID.  (Again just thinking of simplicity.)

I think we should avoid the path because of duplication. If the same widget is rendered twice this would cause an error. Where using uuid even if there are duplicates the waveform would be extracted.

**Comment by Faaatman** [comment]
> I would consider it a misuse of the plugin to simultaneously write to the same path, so the path should be unique among the simultaneous requests.

It is a misuse, but I still think UUID would be the better approach since it's not prone to as many problems as the path approach down the line. And it would stop people from opening issues when they misuse the plugin, instead, it would just make their apps a little bit slower.

**Comment by ryanheise** [comment]
This problem is not a a real problem since it can be documented and an exception can be thrown if the same output file is being used in two simultaneous requests. You mentioned there are more problems, though?

**Comment by ryanheise** [comment]
I'll add to the above that using a UUID does not actually solve the problem of using the same output file in simultaneous requests - an exception would be needed.

**Comment by ryanheise** [comment]
This PR has now been merged and released, and I have done some minor cleanup and removed the uuid key in favour of the output filename. Thanks, @AlexSmirnov9107 for the contribution.
